authentication/ldap_authentication_provider.py

```diff
@@ -12,35 +12,147 @@
+                               'LDAP configuration is invalid.')
+        return who
+
+    def authenticate_by_attribute(self, who, password, default_attr=None):
+        try:
+            conn = self.synchronous_bind(self.admin_user_id,
+                                         self.admin_password)
+        except AuthenticationException as e:
+            err = 'Failed binding to \'{0}\' using the dedicated ' \
+                  'admin_dn: {1}, provided in the configuration; {2}'\
+                .format(self.ldap_url, who, e)
+            self.lgr.warning(err)
+            raise RuntimeError(err)
+
+        attribute_name = self.user_id_attribute or default_attr
+        search_filter = '{0}={1}'.format(attribute_name, who)
+
+        result_list = []
+        try:
+            # Search for matches under the base_dn subtree using the attribute
+            # search filter.
+            ldap_result_id = conn.search(self.base_dn,
+                                         ldap.SCOPE_SUBTREE,
+                                         search_filter,
+                                         [DN_ATTRIBUTE_NAME])
+        except ldap.LDAPError as e:
+            err = 'Failed searching for user under base DN: {0}, according ' \
+                  'to the following attribute filter pattern {1}; {2}'\
+                .format(self.base_dn, search_filter, e)
+            self.handle_authentication_error(err, conn)
+
+        try:
+            # Retrieve query results according to the result ID in order to
+            # obtain the user DN to bind with.
+            while True:
+                result_type, result_data = conn.result(ldap_result_id, 0)
+                if result_data:
+                    if result_type == ldap.RES_SEARCH_ENTRY:
+                        result_list.append(result_data)
+                else:
+                    break
+        except ldap.LDAPError as e:
+            err = 'Failed retrieving results according to result_id; {0}'\
+                .format(e)
+            self.handle_authentication_error(err, conn)
+
+        # unbind admin connection.
+        conn.unbind()
+
+        if len(result_list) != 1:
+            err = 'Expecting a single result matching the user with username' \
+                  ' {0}. Found {1}.'.format(who, len(result_list))
+            self.handle_authentication_error(err, conn)
+
+        # extract the user DN from the search result.
+        user_dn = result_list.pop()[0][0]
+
+        # attempt to bind using the matching user DN and provided password.
+        self.try_bind_s(user_dn, password)
+
+    def handle_authentication_error(self, err, conn):
+        conn.unbind()
+        self.lgr.warning(err)
+        raise AuthenticationException(err)
+
+    def try_bind_s(self, who, password):
+            conn = self.synchronous_bind(who, password)
+            conn.unbind()
+
+    def synchronous_bind(self, who, password):
```

Comment:
since synchronous is the only way we work, maybe no need to name it here. it's a bit confusing in my opinion (yes I am aware of the async API but we're not using it)